File: src/main/java/com/senderman/lastkatkabot/tempobjects/UserRow.java
package com.senderman.lastkatkabot.tempobjects;

import com.annimon.tgbotsmodule.api.methods.Methods;
import com.senderman.TgUser;
import com.senderman.lastkatkabot.Services;
import org.telegram.telegrambots.meta.api.methods.ParseMode;
import org.telegram.telegrambots.meta.api.objects.Message;

import java.util.HashSet;
import java.util.Set;

public class UserRow {
    private final int divider;
    private final String name;
    private long chatId;
    private int messageId;
    private final Set<Integer> checkedUsers;
    private String messageText;

    public UserRow(Message message) throws Exception {
        var lines = message.getText().strip().split("\n");
        if (lines.length != 3)
            throw new Exception("Неверный формат");

        chatId = message.getChatId();
        String title = lines[0].split(" ",2)[1];
        name = lines[1];
        divider = Integer.parseInt(lines[2]);
        checkedUsers = new HashSet<>();
        messageText = "<b>" + title + ":</b>\n\n";
        var resultMessage = Services.handler().sendMessage(message.getChatId(), messageText);
        messageId = resultMessage.getMessageId();
        Services.db().saveRow(chatId, this);
    }

    public void addUser(Message message) {
        var user = new TgUser(message.getFrom());
        if (checkedUsers.contains(user.getId()))
            return;

        checkedUsers.add(user.getId());
        if (checkedUsers.size() % divider == 0)
            messageText += checkedUsers.size() + ". " +user.getLink() + " - " + name + "!\n";
        else
            messageText += checkedUsers.size() + ". " +user.getLink() + " - не " + name + "\n";
        updateMessage();
    }

    private void updateMessage() {
        Methods.editMessageText()
                .setChatId(chatId)
                .setMessageId(messageId)
                .setText(messageText)
                .setParseMode(ParseMode.HTML)
                .call(Services.handler());
        Services.db().saveRow(chatId, this);
    }

    public int getMessageId() {
        return messageId;
    }
}

File: src/main/kotlin/com/senderman/lastkatkabot/tempobjects/UserRow.kt
package com.senderman.lastkatkabot.tempobjects

import com.senderman.lastkatkabot.Services
import com.senderman.neblib.TgUser
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageText
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.User
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException
import java.util.*

class UserRow(message: Message) {
    private val maxUsers = 50
    private val divider: Int
    private val name: String
    private val chatId: Long
    val messageId: Int
    private val checkedUsers: MutableSet<Int>
    private var messageText: String

    init {
        val lines = message.text.trim().split("\n")
        if (lines.size != 3) throw Exception("Неверный формат")
        chatId = message.chatId
        val title = lines[0].split(" ".toRegex(), 2)[1]
        name = lines[1]
        divider = lines[2].toInt()
        if (divider <= 0) throw Exception("Неположительное число")
        checkedUsers = HashSet()
        messageText = "<b>$title:</b>\n\n"
        val resultMessage = Services.handler.sendMessage(message.chatId, messageText)
        messageId = resultMessage.messageId
        Services.db.saveRow(chatId, this)
    }

    fun addUser(newUser: User) {
        if (newUser.id in checkedUsers) return
        if (checkedUsers.size +1 > maxUsers) {
            deleteRow()
            return
        }
        
        val user = TgUser(newUser)
        checkedUsers.add(user.id)
        val pref = if (checkedUsers.size % divider == 0) "" else "не"
        messageText += "${checkedUsers.size}. ${user.link} - $pref $name\n"
        try {
            Services.handler.execute(
                EditMessageText()
                    .setChatId(chatId)
                    .setMessageId(messageId)
                    .setText(messageText)
                    .enableHtml(true)
            )
        } catch (e: TelegramApiRequestException) {
            deleteRow()
        }
        Services.db.saveRow(chatId, this)
    }

    private fun deleteRow(){
        Services.handler.userRows.remove(chatId)
        Services.db.deleteRow(chatId)
    }
}